### PR TITLE
HUB-504 All capture of the AB Testing Variant name

### PR DESCRIPTION
**Pull request by j-maynard**
These changes we hope will allow us to capture the AB Test Variant Name
in the event system which will allow us to to capture this information
even when analytics aren't enabled by the user.

Ticket:
https://govukverify.atlassian.net/secure/RapidBoard.jspa?rapidView=144&modal=detail&selectedIssue=HUB-504

Co-authored-by: John Watts <john.watts@digital.cabinet-office.gov.uk>

### CODE REVIEW

File: shutdown.sh [hunk 1]
```diff
@@ -8,7 +8,8 @@ done
 
 if docker ps | grep hub-redis >/dev/null ; then
     docker stop hub-redis
-    docker rm hub-redis
+    # Docker removes hub-redis when it stops
```
Comment (galund) [inline shutdown.sh:11]: I think this should go in a separate PR. My quick internet searchings suggest that `stop` should not be doing an `rm` so maybe further investigation req'd.

File: hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/AuthnRequestFromTransactionResourceIntegrationTest.java [hunk 2]
```diff
@@ -138,11 +139,37 @@ public void selectIdp_shouldReturnSuccessResponseAndAudit() throws JsonProcessin
...
+    public void selectIdp_shouldRecordAbTestVariant() throws JsonProcessingException {
+        sessionId = aSessionIsCreated();
+        Response response = postIdpSelected(new IdpSelected(idpEntityId, principalIpAddress, REGISTERING, LEVEL_2, "this-is-an-analytics-session-id", "this-is-a-journey-type", abTestVariant));
+
+        assertThat(response.getStatus()).isEqualTo(Response.Status.CREATED.getStatusCode());
+        assertThat(eventSinkStub.getRecordedRequest()).hasSize(2); // one session started event, one idp selected event
+
+        String recordedEvent = new String(eventSinkStub.getLastRequest().getEntityBytes());
+        assertThat(recordedEvent).contains(sessionId.getSessionId());
+        assertThat(recordedEvent).contains(abTestVariant);
```
[the remark this answers is not in the document]
Comment (galund) [reply hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/AuthnRequestFromTransactionResourceIntegrationTest.java:152]: (This is not a requested change, just to be clear - just the other bit about the docker.)

File: shutdown.sh (hunk 1 above)
Comment (j-maynard) [reply shutdown.sh:11]: I am happy to move this into a different PR.  Its all to do with the way Docker is started by other scripts.  All the scripts start docker with `--rm` which tells docker to remove the container once its done.  Whoever made that change didn't update the shutdown scripts to reflect that change.  I'll move this to a new PR but we should probably settle on an approach to these scripts.

File: hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/AuthnRequestFromTransactionResourceIntegrationTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (j-maynard) [reply hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/AuthnRequestFromTransactionResourceIntegrationTest.java:152]: I was following the same pattern used by other tests here to be consistent.

**Comment by j-maynard** [comment]
I've made a minor change so that the code can handle gracefully a missing ABTestVariant field in the incoming JSON.

File: hub/policy/src/main/java/uk/gov/ida/hub/policy/domain/IdpSelectedWithAbTest.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package uk.gov.ida.hub.policy.domain;
```
Comment (jkunle) [inline hub/policy/src/main/java/uk/gov/ida/hub/policy/domain/IdpSelectedWithAbTest.java:1]: I think this might have been slipped in by the IDE

File: hub/policy/src/main/java/uk/gov/ida/hub/policy/domain/state/IdpSelectedState.java [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
```
Comment (jkunle) [inline hub/policy/src/main/java/uk/gov/ida/hub/policy/domain/state/IdpSelectedState.java:5]: Don't need this any more

File: hub/policy/src/test/java/uk/gov/ida/hub/policy/domain/controller/IdpSelectorTest.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
     private static final String IDP_ENTITY_ID = "IDP-a";
     private static final String OTHER_IDP_ENTITY_ID = "idp-b";
     private static final LevelOfAssurance REQUESTED_LOA = LevelOfAssurance.LEVEL_2;
+    private static final String abTestVariant = null;
```
Comment (jkunle) [inline hub/policy/src/test/java/uk/gov/ida/hub/policy/domain/controller/IdpSelectorTest.java:30]: This is no longer used

File: hub/policy/src/test/java/uk/gov/ida/hub/policy/builder/state/IdpSelectedStateBuilder.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
     private SessionId sessionId = SessionIdBuilder.aSessionId().build();
     private List<String> availableIdentityProviders = List.of("idp-a", "idp-b", "idp-c");
     private boolean transactionSupportsEidas = false;
+    private String abTestVariant;
```
Comment (jkunle) [inline hub/policy/src/test/java/uk/gov/ida/hub/policy/builder/state/IdpSelectedStateBuilder.java:30]: No longer needed

Comment (jkunle) [reply hub/policy/src/test/java/uk/gov/ida/hub/policy/builder/state/IdpSelectedStateBuilder.java:30]: 👍 It's all looking good to me just a few things above.

Comment (j-maynard) [reply hub/policy/src/test/java/uk/gov/ida/hub/policy/builder/state/IdpSelectedStateBuilder.java:30]: This is used in the file.

File: hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/SessionResourceWithRedisIntegrationTest.java [hunk 7]
```diff
@@ -60,6 +60,7 @@
     private static final boolean REGISTERING = true;
     private static final boolean SIGNING_IN = !REGISTERING;
     private static final LevelOfAssurance REQUESTED_LOA = LevelOfAssurance.LEVEL_2;
+    private static final String abTestVariant = null;
```
Comment (galund) [inline hub/policy/src/integration-test/java/uk/gov/ida/integrationtest/hub/policy/apprule/SessionResourceWithRedisIntegrationTest.java:63]: I feel like in these cases it would have been better to stick to the existing pattern i.e. just pass "this-is-a-test-variant-specifier" or something, rather than have a pointless `null` kicking around.
